src/Tokenizer/Tokens.php

```diff
@@ -1102,7 +1098,7 @@ public function toJson(): string
     public function isAllTokenKindsFound(array $tokenKinds): bool
     {
         foreach ($tokenKinds as $tokenKind) {
-            if (!isset($this->foundTokenKinds[$tokenKind])) {
+            if (0 === ($this->foundTokenKinds[$tokenKind] ?? 0)) {
```

Comment:
Do I understand correctly we check if the index exists _and_ value under that index is _not null_? Maybe we should do that more explicitly (`array_key_exists()` + `null !== value`)?

Reply:
We check if the index does not exist or the value is zero.

the condition is clear - first we do `$arr[$k] ?? 0` and then `expr === 0`, which is also the fastest

Reply:
I understand how it works, but the actual logic is hidden behind implementation. You could use `if ('non-existent' === ($this->foundTokenKinds[$tokenKind] ?? 'non-existent'))` and it would work the same, because `0` is only a helper value used as a fallback and then compared.

`if (!array_key_exists($tokenKind, $this->foundTokenKinds) || null === $this->foundTokenKinds[$tokenKind])` is only slightly longer but shows the intent much better IMHO. But it's cosmetics, I won't force it.

Reply:
I prefer the current implementation which does the lookup only once. It is quite performance critical method.